File: pe/nonvisualstudio/haskell/Problem0028.hs
-- ghci
-- :load C:\Users\Thomas\Documents\GitHub\practice\pe\nonvisualstudio\haskell\Problem0028.hs
-- :r
-- :set +s for times

module Problem0028 where

  sumOfBoxLevel n = sumOfBoxLevelRecur 0 0
    where
      nextNumberToAdd i = ((n * n) - ((n - 1) * i))
      sumOfBoxLevelRecur i result | i == 4     = result
                                  | otherwise  = sumOfBoxLevelRecur (i + 1) (result + (nextNumberToAdd i))


  sumOfSpiralDiagonals n = sumOfSpiralDiagonalsRecur n 1
    where
      nextNumberToAdd i = sumOfBoxLevel i
      sumOfSpiralDiagonalsRecur i result | i == 1     = result
                                         | otherwise  = sumOfSpiralDiagonalsRecur (i - 2) (result + (nextNumberToAdd i))


  -- tests

  sumOfSpiralDiagonalsTests = and
    [
      sumOfSpiralDiagonals 1 == 1,
      sumOfSpiralDiagonals 5 == 101,
      sumOfSpiralDiagonals 3 == 25
    ]

  sumOfBoxLevelTests = and
    [
      sumOfBoxLevel 3 == 24,
      sumOfBoxLevel 5 == 76
    ]

  tests = and
		[
			sumOfBoxLevelTests,
      sumOfSpiralDiagonalsTests
		]


  answer = sumOfSpiralDiagonals 1001

File: pe/visualstudio/PE.FSharp/Problem0028.fs
namespace PE.FSharp

module Problem0028 =
    let sumOfBoxCorners n = 
        let rec sumOfBoxCornersRecur i result =
            let nextNumberToAdd = ((n * n) - ((n - 1) * i))
                                   
            match i with
            | 4 -> result
            | _ -> sumOfBoxCornersRecur (i + 1) (result + nextNumberToAdd)

        sumOfBoxCornersRecur 0 0

    let sumOfSpiralDiagonals n =
        let rec sumOfSpiralDiagonalsRecur i result =
            match i with
            | 1 -> result
            | _ -> sumOfSpiralDiagonalsRecur (i - 2) (result + sumOfBoxCorners i)

        sumOfSpiralDiagonalsRecur n 1